lib/ecmascript.ts

```diff
@@ -3232,444 +3360,468 @@ export const ES = ObjectAssign({}, ES2020, {
-  },
-  RoundTime: (
+    hours,
+    minutes,
+    seconds,
+    milliseconds,
+    microseconds,
+    nanoseconds
+  ));
+  days += deltaDays;
+
+  // Delegate the date part addition to the calendar
+  const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
+  const datePart = CreateTemporalDate(year, month, day, calendar);
+  const dateDuration = new TemporalDuration(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
+  const addedDate = CalendarDateAdd(calendar, datePart, dateDuration, options);
+
+  return {
+    year: GetSlot(addedDate, ISO_YEAR),
+    month: GetSlot(addedDate, ISO_MONTH),
+    day: GetSlot(addedDate, ISO_DAY),
+    hour,
+    minute,
+    second,
+    millisecond,
+    microsecond,
+    nanosecond
+  };
+}
+
+export function AddZonedDateTime(
+  instant,
+  timeZone,
+  calendar,
+  years,
+  months,
+  weeks,
+  days,
+  h,
+  min,
+  s,
+  ms,
+  µs,
+  ns,
+  options?: any
+) {
+  // If only time is to be added, then use Instant math. It's not OK to fall
+  // through to the date/time code below because compatible disambiguation in
+  // the PlainDateTime=>Instant conversion will change the offset of any
+  // ZonedDateTime in the repeated clock time after a backwards transition.
+  // When adding/subtracting time units and not dates, this disambiguation is
+  // not expected and so is avoided below via a fast path for time-only
+  // arithmetic.
+  // BTW, this behavior is similar in spirit to offset: 'prefer' in `with`.
+  const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
+  if (DurationSign(years, months, weeks, days, 0, 0, 0, 0, 0, 0) === 0) {
+    return AddInstant(GetSlot(instant, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+  }
+
+  // RFC 5545 requires the date portion to be added in calendar days and the
+  // time portion to be added in exact time.
+  const dt = BuiltinTimeZoneGetPlainDateTimeFor(timeZone, instant, calendar);
+  const datePart = CreateTemporalDate(GetSlot(dt, ISO_YEAR), GetSlot(dt, ISO_MONTH), GetSlot(dt, ISO_DAY), calendar);
+  const dateDuration = new TemporalDuration(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
+  const addedDate = CalendarDateAdd(calendar, datePart, dateDuration, options);
+  const dtIntermediate = CreateTemporalDateTime(
+    GetSlot(addedDate, ISO_YEAR),
+    GetSlot(addedDate, ISO_MONTH),
+    GetSlot(addedDate, ISO_DAY),
+    GetSlot(dt, ISO_HOUR),
+    GetSlot(dt, ISO_MINUTE),
+    GetSlot(dt, ISO_SECOND),
+    GetSlot(dt, ISO_MILLISECOND),
+    GetSlot(dt, ISO_MICROSECOND),
+    GetSlot(dt, ISO_NANOSECOND),
+    calendar
+  );
+
+  // Note that 'compatible' is used below because this disambiguation behavior
+  // is required by RFC 5545.
+  const instantIntermediate = BuiltinTimeZoneGetInstantFor(timeZone, dtIntermediate, 'compatible');
+  return AddInstant(GetSlot(instantIntermediate, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+}
+
+export function RoundNumberToIncrement(quantity, increment, mode) {
+  if (increment === 1) return quantity;
```

Comment:
From a quick look at usage of this function, it seems like increment is usually a Number, not a bigint. Also, `1n === 1` is `false`, so if the inputs are bigint then this test will never succeed. So I think it's good to type `increment` as a Number and `quantity` is a BigInt.

Reply:
> usually a Number

FWIW, I think "is always a Number except for bugs" is probably more accurate. @ptomato may have more insight.

Reply:
That may be the case, but it probably _ought_ to be a BigInt. The big-integer wrapper library we use allows sloppy mixing of BigInt and Number types, where real BigInts wouldn't.

Reply:
FWIW, JSBI's typings doesn't allow this (at least, for some of the functions), so deciding on one or the other and using it consistently will be better enforced as that migration + better typings are landed throughout.

It seems that they should both be bigints?

Reply:
> It seems that they should both be bigints?

I dug into where `increment` is used. It's clear that `increment` (which often comes from user-entered Number-typed options) is always a Number. I found only one exception which looks like it's caused by a bug in the ZonedDateTime implementation. See  https://github.com/js-temporal/temporal-polyfill/pull/57#discussion_r718042170. @12wrigja are there others?

Changing all usage of `increment` (~60 lines) to `bigint` seems like a pretty significant refactor with non-trivial impact on the types of other variables and function parameters. I'd be hesitant to do it before this PR is merged and before we have non-`any` types on parameters and return types of ES methods, because having everything typed would make it much easier to understand downstream impact and avoid mistakes.  I'd also like to understand the bundle-size impact of a big refactor like that, esp. in the case of `increment` where AFAICT always smaller than MAX_SAFE_INTEGER. 

For this PR (or a separate PR before merging this one), my suggestion is that we leave usage as Number and do the minimal required to fix known type bugs, e.g. `ZonedDateTime.p.round()` should coerce `dayLengthNs` to Number before passing to `RoundISODateTime`. 

Then we can discuss as a follow-up whether (and if so, when) we want to change `increment` (and other similar parameters) to bigint globally.

Would that work?

> > The big-integer wrapper library we use allows sloppy mixing of BigInt and Number types, where real BigInts wouldn't.
>
> JSBI's typings doesn't allow this (at least, for some of the functions), so deciding on one or the other and using it consistently will be better enforced as that migration + better typings are landed throughout.

Stricter types sounds really helpful. Even if our bigint library allows passing `number` params to some methods, we could always wrap it with stricter types ourselves. But I really like the idea of not doing any library math between bigint and integer values if the native operators wouldn't allow it. 

Not sure if making that change would blow up the complexity of this PR though. I'd leave it up to @12wrigja which order is best.

Reply:
> I found only one exception which looks like it's caused by a bug in the ZonedDateTime implementation. See #57 (comment). @12wrigja are there others?
That was the only one I found as well.

> For this PR (or a separate PR before merging this one), my suggestion is that we leave usage as Number and do the minimal required to fix known type bugs, e.g. ZonedDateTime.p.round() should coerce dayLengthNs to Number before passing to RoundISODateTime.

SGTM, I'll definitely be following up on this and other similar issues once I return to integrating JSBI (hopefully today / later this week).